### PR TITLE
Authors

### PR DESCRIPTION
**Pull request by amymhaddad**
@paulghaddad Here's what's included in this PR:

In app.py:
Updated `@app.route("/authors", methods=["GET", "POST"])` to include:
-Data from database and database query
-Sorting 
-Filtering
-Pagination

Updated `@app.route("/authors/<int:author_id>")` to include:
-Data from database and database query

Updated `authors.html` and `main.css`:
-To display and format the sorting, filtering, and pagination functions

Updated `author.html` and `main.css`:
-To display and format data 

Updated `books.html` and `main.css`:
-I ran a new script to get the updated books data imported into the database. Then, I styled the data with CSS. 

***The authors and author pages DO NOT show the book images. I need to join this table with the books table so I can get the correct book to display for each author. I'll open a new PR for that.

### CODE REVIEW

File: alembic/versions/7b206ee27574_adding_authors_table.py [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+
+def upgrade():
+    op.alter_column('books', 'image_url',
+               existing_type=sa.VARCHAR(),
+               nullable=True)
+    op.alter_column('books', 'summary',
+               existing_type=sa.VARCHAR(),
+               nullable=True)
+
```
Comment (amymhaddad) [inline alembic/versions/7b206ee27574_adding_authors_table.py]: Three lines :(

File: alembic/versions/9e475da39630_add_authors_table.py [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+depends_on = None
+
+
+def upgrade():
+    op.create_table(
+        "authors",
+        sa.Column("id", sa.Integer(), nullable=False),
+        sa.Column("first_name", sa.String(), nullable=False),
+        sa.Column("last_name", sa.String(), nullable=False),
+        sa.Column("biography", sa.String(), nullable=True),
```
Comment (amymhaddad) [inline alembic/versions/9e475da39630_add_authors_table.py:23]: 👍

File: app.py [hunk 3]
```diff
@@ -48,10 +48,27 @@ def book(book_id):
...
     """Show the list of authors and the books they've written"""
-    return render_template("authors.html", authors=authors_data)
+
+    if request.method == "POST":
+        single_author = request.form["last_name"]
+        authors = db_session.query(Author).filter(Author.last_name == single_author)
+
+    elif request.method == "GET" and request.args:
+        if request.args.get("page"):
+            select_authors = request.args.get("page")
```
Comment (amymhaddad) [inline app.py]: `select_authors` is not a good name...make it more descriptive

File: app.py [hunk 4]
```diff
@@ -48,10 +48,27 @@ def book(book_id):
...
+            authors = db_session.query(Author).limit(10).offset(select_authors)
+
+        elif request.args.get("sort"):
+            name = request.args.get("sort")
```
Comment (amymhaddad) [inline app.py]: Ditto

File: app.py [hunk 5]
```diff
@@ -83,12 +100,9 @@ def category(category_id):
 def author(author_id):
     """Display a page about an author"""
 
-    author_object = None
-    for author in authors_data:
-        if author["author_id"] == author_id:
-            author_object = author
+    author = db_session.query(Author).filter(Author.id == author_id).one()
```
Comment (amymhaddad) [inline app.py:103]: Good

File: models/author.py [hunk 6]
```diff
@@ -0,0 +1,12 @@
+from sqlalchemy import Column, Integer, Numeric, String, ForeignKey
```
Comment (amymhaddad) [inline models/author.py]: You're not using all these imports

File: models/author.py [hunk 7]
```diff
@@ -0,0 +1,12 @@
...
+from sqlalchemy.orm import relationship
```
Comment (amymhaddad) [inline models/author.py]: Ditto

File: scripts/import_authors.py [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+import random
+
+from upcase_booklist_app.database import db_session
+from upcase_booklist_app.models.author import Author
+
+session = db_session()
+
+authors = []
+
+for _ in range(100):
```
Comment (amymhaddad) [inline scripts/import_authors.py:13]: Really nice

File: static/main.css [hunk 9]
```diff
@@ -159,6 +150,43 @@
...
+.sort-authors {
+    display: flex;
+    flex-direction: row;
+    margin-right: 10px;
+    text-decoration: none;
+    padding-bottom: 20px;
+    padding-top: 8px;
+}
+
+.sort-authors a{
```
Comment (amymhaddad) [inline static/main.css]: space before {

File: templates/author.html [hunk 10]
```diff
@@ -17,29 +17,11 @@
...
-                                <img src="{{ book['image'] }}" alt="{{ book['title'] }}">
-                            </a>
-                        <div>
-                            <p class="space book-title">{{ book['title'] }}</p>
-                            <p class="space text">{{ book['summary'] }}</p>
-                        </div>  
-
-                    </div>  
-                {% endfor %}
+                    <h1>{{ author.first_name }} {{ author.last_name}}</h1>
```
Comment (amymhaddad) [inline templates/author.html]: Consider making a `@property` on your `Author` model for `display_name` or `full_name`, which is very useful and can be calculated from the two attributes above (so doesn't need to be in the db).

File: templates/authors.html [hunk 11]
```diff
@@ -19,28 +19,32 @@
...
-                    
-                    {% for book in author['books'] %}
-                        <div class="book">
-                            <a href="{{ url_for('book', book_id=book['book_id']) }}">
-                                <img src="{{ book['image'] }}" alt="{{book['title'] }}">
-                            </a>
-                        </div>
-                    
-                    </div>
+                        <a href="{{ url_for('author', author_id=author.id) }}">{{ author.first_name +' ' +author.last_name }}</a>
```
Comment (amymhaddad) [inline templates/authors.html]: `full_name` would simplify this line here

**Review by amymhaddad — COMMENTED** [review]
Mostly good; nice work 💯

**Comment by amymhaddad** [comment]
@paulghaddad Great suggestion on using @property! Here are the changes I made:

In `app.py`:
-Updated variable names (limit_authors and author_name) in this code:

```
if request.args.get("page"):
            limit_authors = request.args.get("page")
            authors = db_session.query(Author).limit(10).offset(limit_authors)

        elif request.args.get("sort"):
            author_name = request.args.get("sort")
            authors = db_session.query(Author).order_by(author_name).all()
```

In `author.py`:
-Added @property decorator 
--->However, my alembic migration shows as empty. Is that normal?

@property
    def display_name(self):
        return self.first_name + ' ' + self.last_name

In `authors.html `and `author.html`:
-Use the @property deocorator in the html itself

***Please check my alembic migration (begins: 411bc...). It shows up as empty and I'm not sure why.

File: alembic/versions/411bcc6b39b8_adding_property_to_author_table.py [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+
+# revision identifiers, used by Alembic.
+revision = '411bcc6b39b8'
+down_revision = '9e475da39630'
+branch_labels = None
+depends_on = None
+
+
+def upgrade():
+    # ### commands auto generated by Alembic - please adjust! ###
```
Comment (paulghaddad) [inline alembic/versions/411bcc6b39b8_adding_property_to_author_table.py]: You still have an empty migration. I'll approve but make a subsequent PR to remove it.

File: models/author.py [hunk 13]
```diff
@@ -0,0 +1,15 @@
...
+
+    __tablename__ = "authors"
+
+    id = Column(Integer, primary_key=True)
+    first_name = Column(String(), nullable=False)
+    last_name = Column(String(), nullable=False)
+    biography = Column(String(), nullable=True)
+
+    @property
+    def display_name(self):
```
Comment (paulghaddad) [inline models/author.py:14]: 👍 although I'd personally use an f-string.

File: templates/author.html [hunk 14]
```diff
@@ -17,29 +17,11 @@
...
-                                <img src="{{ book['image'] }}" alt="{{ book['title'] }}">
-                            </a>
-                        <div>
-                            <p class="space book-title">{{ book['title'] }}</p>
-                            <p class="space text">{{ book['summary'] }}</p>
-                        </div>  
-
-                    </div>  
-                {% endfor %}
+                    <h1>{{ author.display_name }}</h1>
```
Comment (paulghaddad) [inline templates/author.html:23]: Much simpler :)

**Review by paulghaddad — APPROVED** [review]
Really nice job on this. Per my comment, please make another PR to remove the empty migration, otherwise awesome! 💯

**Pull request merged**